controller/lib/core/flow_integrator.cpp

```diff
@@ -14,63 +14,78 @@ limitations under the License.
 
 void FlowIntegrator::AddFlow(Time now, VolumetricFlow uncorrected_flow) {
-  VolumetricFlow flow = uncorrected_flow + flow_correction_;
-  Duration delta = now - last_flow_measurement_time_;
-  if (delta >= VOLUME_INTEGRAL_INTERVAL) {
-    volume_ += delta * (last_flow_ + flow) / 2.0f;
+  VolumetricFlow flow = uncorrected_flow + FlowCorrection();
+
+  // First time?  In that case, don't need to adjust the flow.  Just record
+  // this data point.
+  if (last_flow_measurement_time_ == std::nullopt) {
+    last_note_time_ = now;
+    last_note_vol_ = ml(0);
+    last_flow_measurement_time_ = now;
+    last_flow_ = flow;
+    return;
+  }
+
+  Duration dt = now - *last_flow_measurement_time_;
+  if (dt >= VOLUME_INTEGRAL_INTERVAL) {
+    volume_ += dt * (last_flow_ + flow) / 2.0f;
     last_flow_measurement_time_ = now;
     last_flow_ = flow;
   }
 }
 
 void FlowIntegrator::NoteExpectedVolume(Volume v) {
-  flow_correction_ = ml_per_sec(flow_correction_pid_.Compute(
-      last_flow_measurement_time_, volume_.ml(), v.ml()));
+  // We need at least one call to AddFlow before we can do anything.
+  if (last_note_time_ == std::nullopt ||
+      last_flow_measurement_time_ == std::nullopt ||
+      *last_note_time_ >= *last_flow_measurement_time_) {
+    return;
+  }
+
+  // Compute the new flow correction as follows.
+  //
+  // - Compute the time delta between now (last_flow_measurement_time_) and the
+  //   last call to NoteExpectedVolume.
+  Duration dt = *last_flow_measurement_time_ - *last_note_time_;
+
+  // - Compute be what our volume would have been if we hadn't applied any
+  //   flow correction during the last breath.
+  Volume v_uncorrected = volume_ - flow_correction_ * dt;
+
+  // - Compute the difference in volume between now and the previous breath
+  //   we would have seen if we hadn't applied correction.
+  Volume dv = v_uncorrected - last_note_vol_;
+
+  // - Assume that dv will remain constant.  Then we can compute the expected
+  //   volume after the next breath.
+  Volume v_next = volume_ + dv;
+
+  // - Finally, choose a flow correction that drives v_next to v, assuming
+  //   dt remains constant.
+  flow_correction_ = (v - v_next) / dt;
+
+  // NOTE: You might be tempted to apply some historesis to dv, e.g. using an
+  // exponentially-weighted average.  Experimenting with the data checked in
+  // under sample-data suggests this doesn't help.  dv changes by as much as
+  // 15ml/s between breaths, but at least from my eyeballing of the data, using
+  // multiple historical values of dv makes our prediction of the next dv
+  // worse.
+  //
+  // TODO: That said, this algorithm handles coughing, stacked breaths, etc
+  // very badly, treating basically anything other than a completely uniform
+  // breath as flow error, and aggressively "correcting" it.
+  //
+  // TODO: Similarly, if the time until the next breath happens to be much
+  // longer than dt, we may seriously overcorrect the volume.
```

Comment:
Does this algorithm have the property of being able to eventually recover from such errors if things proceed smoothly afterwards?

Reply:
Yes, and I observe this because on the first breath my test lung behaves differently than on all other breaths, and we recover from this no prob.  If anything it will recover "too quickly", masking the fact that these occasions (which may be clinically significant) occurred.